### ISSUE TITLE
Provide built-in to do conversion between different string encoding types (base64, hex, etc)

### ISSUE THREAD

**Issue by me-viper** [open-policy-agent/opa#7198]
## What is the underlying problem you're trying to solve?

When dealing with web stuff it's common to receive non-string inputs as base64 encoded strings. On the other hand, opa built-ins convert non-string stuff into hex strings. This creates issues when comparison is needed, for example:

```
example if {
    sig := sprintf("%s %s%s %s", [input.method, input.path, input.queryString, input.protocol])
    hash := crypto.hmac.sha256(sig, data.nbo.key) # hex encoded string
    signature := input.signature # base64 endoded string
    # Can't do that. It's possible to convert both to string and do the comparison but that is asking for trouble.
    crypto.hmac.equal(hash, signature)
}
```

## Describe the ideal solution
Some built-in(s) that enables to do conversion between hex encoded strings and base64 encoded strings.

```
hex.to_base64(s)
hex.to_base64url(s)
base64.to_hex(s)
base64.to_base64url(s)
...
```

Maybe something more generic like:

```
convert_encoing(s, source_type, d, destination_type)
where type: base64, base64url or hex
```

Or something that returns byte array:

```
hex.bytes(s)
base64.bytes(s)
base64url.bytes(s)
```

## Describe a "Good Enough" solution

N/A

## Additional Context

N/A

**Comment by srenatus** [comment]
There's another avenue worth exploring: If you're using Wasm, and your Wasm SDK allows you to provide "SDK-dependent" builtins, you can
1. create your own capabilities file, adding the type signature of a builtin you'd like to have
2. use `opa built -t wasm --capabilities caps.json ...` to build the Wasm module
3. use the Wasm module with the described builtin provided through your SDK.

It's probably the quickest way to get what you want, compared to adding the new builtins as native and Wasm implementations.... Worth trying, at least.

**Comment by me-viper** [comment]
Yes, we're already doing that. It works, writing tests for policies require extra hoops though.

But it would be nice to have it out of box since base64 inputs are very common in web scenarios.
I can do a PR if I get an input on how these built-ins should look like.

**Comment by me-viper** [comment]
Personally, I'd go byte conversion way:

```
string hex._from_bytes(byte[])
byte[] hex.to_bytes(string)
...
base64.from_bytes()
base64.to_bytes()
...
```
Kinda low level but in the end, we're dealing with bytes anyway.

**Comment by anderseknert** [comment]
From the issue description, I'm thinking we could just have `crypto.hmac.equal` accept either hex or base64 encoding for any of its arguments?

**Comment by me-viper** [comment]
Yes. It will be the minimum that works.

**Comment by anderseknert** [comment]
👍 I wouldn't mind that change... you see any issues with that @srenatus?
